pysages/ssages/ffs.py

```diff
@@ -0,0 +1,181 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+# Copyright (c) 2020: SSAGES Team (see LICENSE.md)
+
+from collections import namedtuple
+from jax import scipy
+from pysages.ssages.grids import build_indexer
+
+from .core import GriddedSamplingMethod, generalize  # pylint: disable=relative-beyond-top-level
+
+import jax.numpy as np
+
+# ================================== #
+#        Forward Flux Sampling       #
+# ================================== #
+
+class FFSState(
+    namedtuple(
+        "FFSState",
+        ("Phi_A", "Previous_Window", "Current_Window", "Prob_window"),
+    )
+):
+    def __repr__(self):
+        return repr("PySAGES " + type(self).__name__)
+
+#FFS needs a 1D grid in Collective Variable to define the path
+class FFS(GriddedSamplingMethod):
+    def __call__(self, snapshot, helpers):
+        #Number of parallel replicas per window
+        Nmax_replicas= np.asarray(self.kwargs.get('M', 2000))
+        N0_steps= np.asarray(self.kwargs.get('M', 2000))
+        sampling_time= np.asarray(self.kwargs.get('M', 2000))
+        system=0
+        run=0
+        return _ffs(snapshot, self.cv, self.grid, Nmax_replicas, N0_steps,sampling_time,system,run,helpers)
+
+
+def _ffs(snapshot, cv, grid, Nmax_replicas, N0_steps,sampling_time,system,run,helpers):
+    dt = snapshot.dt
+    dims = grid.shape.size
+    N_windows=len(grid)
+    indices, momenta = helpers.indices, helpers.momenta
```

Comment:
We now have a restore method as part of the helpers

```suggestion
    indices, momenta, restore = helpers
```